test/coreputation/upsert_validator.js

```diff
@@ -0,0 +1,127 @@
+      insertedValidator.status.toString(10),
+      `Expected validator status is ${ValidatorStatus.Staked} but found ${insertedValidator.status} `,
+    );
+    assert.strictEqual(
+      inputValidatorInfo.reputation.toString(10),
+      insertedValidator.reputation.toString(10),
+      `Expected validator reputation is ${inputValidatorInfo.reputation} but found ${insertedValidator.reputation} `,
+    );
+  });
+
+  it('should add validator in Deregistered state when his reputation is zero', async () => {
+    await coreputationInstance.upsertValidator(
+      inputValidatorInfo.validator,
+      new BN(0),
+      { from: coconsensus },
+    );
+    const insertedValidator = await coreputationInstance.validators.call(inputValidatorInfo.validator);
+
+    assert.strictEqual(
+      ValidatorStatus.Deregistered.toString(),
+      insertedValidator.status.toString(10),
+      `Expected validator status is ${ValidatorStatus.Deregistered} but found ${insertedValidator.status} `,
+    );
+    assert.strictEqual(
+      (new BN(0)).toString(10),
+      insertedValidator.reputation.toString(10),
+      `Expected validator reputation is ${(new BN(0)).toString(10)} but found ${insertedValidator.reputation} `,
+    );
+  });
+
+  it('should update validator status to Deregistered when reputation is zero', async () => {
+    await coreputationInstance.upsertValidator(
+      inputValidatorInfo.validator,
+      inputValidatorInfo.reputation,
+    );
+    await coreputationInstance.upsertValidator(
+      inputValidatorInfo.validator,
+      new BN(0),
+    );
+    const updatedValidator = await coreputationInstance.validators.call(inputValidatorInfo.validator);
+    assert.strictEqual(
+      ValidatorStatus.Deregistered.toString(),
+      updatedValidator.status.toString(10),
+      `Expected validator status is ${ValidatorStatus.Deregistered} but found ${updatedValidator.status} `,
+    );
+
+    assert.strictEqual(
+      '0',
+      updatedValidator.reputation.toString(10),
+      `Expected validator reputation is 0 but found ${updatedValidator.reputation} `,
+    );
+  });
+
+  it('should not update validator information in slashed state', async () => {
+    await coreputationInstance.upsertValidator(
+      inputValidatorInfo.validator,
+      inputValidatorInfo.reputation,
+    );
+    await coreputationInstance.setValidatorSlashed(
+      inputValidatorInfo.validator,
+    );
+    await coreputationInstance.upsertValidator(
+      inputValidatorInfo.validator,
+      new BN(20),
+    );
+
+    const updatedValidator = await coreputationInstance.validators.call(inputValidatorInfo.validator);
+    assert.strictEqual(
+      ValidatorStatus.Slashed.toString(),
+      updatedValidator.status.toString(10),
+      `Expected validator status is ${ValidatorStatus.Slashed} but found ${updatedValidator.status} `,
+    );
+
+    assert.strictEqual(
+      (new BN(0)).toString(10),
+      updatedValidator.reputation.toString(10),
```

Comment:
For all the tests, the numbers are converted to a string. Why big numbers are not compared directly?

Reply:
I find it very useful when asserts are mismatched and in the messages it clearly shows the expected value Vs matching value. e.g. below BN exact number is difficult to know without converting to string. 
`expected <BN: a124b> to equal 10 `

Also I am not sure if there is an agreement to use only BigNumber comparisons in test assertions.